apstools/devices/motor_factory.py

```diff
@@ -0,0 +1,136 @@
+"""
+Motor Bundle Factory
+====================
+
+Factory for creating bundles with any number of positioners.
+
+.. autosummary::
+
+    ~axis_component
+    ~mb_class_factory
+    ~mb_creator
+
+.. rubric:: Examples
+
+.. code-block:: python
+    :linenos:
+
+    xy_stage = mb_creator(prefix="IOC:", motors={"x": "m21", "y": "m22"})
+
+.. seealso:: :doc:`How to use 'mb_creator()' </examples/ho_motor_factory>`
+"""
+
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Type
+from typing import Union
+
+from ophyd import Component
+from ophyd import EpicsMotor
+from ophyd import Kind
+from ophyd import MotorBundle
+from ophyd import SoftPositioner
+
+from apstools.utils import dynamic_import
+
+DEFAULT_DEVICE_BASE_CLASSES: List[Union[str, Type]] = [MotorBundle]
+DEFAULT_DEVICE_CLASS_NAME: str = "MB_Device"
+DEFAULT_MOTOR_LABELS: List[str] = ["motors"]
+DEFAULT_KIND: Kind = Kind.hinted | Kind.normal
+
+MOTORS_TYPE = Union[List[str], Dict[str, Optional[Union[str, Dict]]]]
+
+
+def axis_component(parms: dict[str, Any]) -> Component:
+    """Return a Component with 'parms for this axis."""
+    parms = parms or {}
+    if isinstance(parms, str):
+        parms = dict(prefix=parms)
+    attrs = dict(kind=DEFAULT_KIND, labels=DEFAULT_MOTOR_LABELS)
+    has_prefix: bool = parms.get("prefix") is not None
+    axis_class_name = parms.pop("class", None)
+    factory = parms.pop("factory", None)
+    if axis_class_name is not None and factory is not None:
+        raise ValueError(f"Unsupported configuration: {axis_class_name=!r} and {factory=!r}")
+    if factory is not None:
+        creator = factory.pop("function")
+        if isinstance(creator, str):
+            creator = dynamic_import(creator)
+        base = creator(**factory)
+        # Add 'prefix' if missing?
+    elif axis_class_name is None:
+        base = EpicsMotor if has_prefix else SoftPositioner
+        if base == EpicsMotor and "prefix" not in parms:
+            parms["prefix"] = ""
+    else:
+        # Allow for any positioner type.
+        base = dynamic_import(axis_class_name)
+    attrs.update(**parms)
+
+    args = [base]
+    prefix = attrs.pop("prefix", None)
+    if prefix is not None:
+        args.append(prefix)
+    return Component(*args, **attrs)
+
+
+def mb_class_factory(
+    motors: MOTORS_TYPE = None,
+    class_bases: List[Union[str, Type]] = DEFAULT_DEVICE_BASE_CLASSES,
+    class_name: str = DEFAULT_DEVICE_CLASS_NAME,
+) -> Type:
+    """Create a custom MotorBundle (or as specified in 'class_bases') class."""
+    motors: MOTORS_TYPE = motors or {}
+    if isinstance(motors, list):
+        motors = {axis: {} for axis in motors}
+
+    if not isinstance(class_bases, list):
+        raise TypeError(f"Must be a list, received {class_bases=!r}")
```

Comment:
This check for classes_base being a list seems unnecessary.

Why must it be a list? Line 95 just needs it to be iterable, and will raise an exception if it's not. Perhaps you can get rid of lines 89-90 and replace the type-hint on line 81 with `collections.abc.Sequence`?

Reply:
The exception that will be raised on line 95 doesn't cut it?

Reply:
exception on line 95?

Reply:
`for base in class_bases` will raise an exception if `class_bases` is not iterable.

Is the `isinstance` check meant to guard against someone passing in e.g. a string, which is iterable but the elements of the sequence are not valid? If that's the case, then don't you actually want to check the elements `class_bases` instead of `class_bases` itself?

Reply:
I'm imagining someone not getting the YAML quite right for this part.  This check seems friendlier than letting the `dynamic_import()` function, having received the first letter of the one base class provided, declare that a dotted name is required.  The might have provided a dotted name but if it was not a list of str, they'd be very confused.  I'll make the class_base a dotted name in this test case: https://github.com/BCDA-APS/apstools/blob/c5d7cd6c86ccdaad0caf288d8242099015e51438/apstools/devices/tests/test_motor_factory.py#L103-L107